File: web/src/lib/filters.test.ts
import { describe, expect, it } from 'vitest';
import {
  countActiveFilters,
  DEFAULT_FILTERS,
  hasActiveFilters,
  matchesFilter,
  type CardFilterState,
} from './filters.ts';
import type { Card } from './types.ts';

const sampleCard = (overrides: Partial<Card> = {}): Card => ({
  id: 'c1',
  day: '2026-08-19',
  title: 'Test Kartı',
  note: 'Açıklama',
  startTime: '10:00',
  endTime: '11:00',
  color: 'red',
  done: false,
  sortIndex: 600,
  manualSort: false,
  habitId: null,
  checklist: [],
  priority: 'none',
  deadlineAt: null,
  tags: [],
  reminders: [],
  images: [],
  createdAt: '2026-08-19T08:00:00Z',
  updatedAt: '2026-08-19T08:00:00Z',
  ...overrides,
});

describe('Card Filters Logic', () => {
  it('varsayılan filtrelerde tüm kartlar eşleşir ve aktif filtre sayısı 0 döner', () => {
    const card = sampleCard();
    expect(matchesFilter(card, DEFAULT_FILTERS)).toBe(true);
    expect(countActiveFilters(DEFAULT_FILTERS)).toBe(0);
    expect(hasActiveFilters(DEFAULT_FILTERS)).toBe(false);
  });

  it('tamamlanma durumuna göre doğru filtreler', () => {
    const todoCard = sampleCard({ done: false });
    const doneCard = sampleCard({ done: true });

    const todoFilter: CardFilterState = { ...DEFAULT_FILTERS, status: 'todo' };
    const doneFilter: CardFilterState = { ...DEFAULT_FILTERS, status: 'done' };

    expect(matchesFilter(todoCard, todoFilter)).toBe(true);
    expect(matchesFilter(doneCard, todoFilter)).toBe(false);

    expect(matchesFilter(todoCard, doneFilter)).toBe(false);
    expect(matchesFilter(doneCard, doneFilter)).toBe(true);
  });

  it('öncelik değerine göre filtreler', () => {
    const urgentCard = sampleCard({ priority: 'urgent' });
    const lowCard = sampleCard({ priority: 'low' });

    const urgentFilter: CardFilterState = { ...DEFAULT_FILTERS, priority: 'urgent' };
    expect(matchesFilter(urgentCard, urgentFilter)).toBe(true);
    expect(matchesFilter(lowCard, urgentFilter)).toBe(false);
  });

  it('etiketlere göre filtreler (büyük/küçük harf duyarsız)', () => {
    const workCard = sampleCard({ tags: ['İş', 'Acil'] });
    const personalCard = sampleCard({ tags: ['Kişisel'] });

    const tagFilter: CardFilterState = { ...DEFAULT_FILTERS, tags: ['iş'] };
    expect(matchesFilter(workCard, tagFilter)).toBe(true);
    expect(matchesFilter(personalCard, tagFilter)).toBe(false);

    const multiTagFilter: CardFilterState = { ...DEFAULT_FILTERS, tags: ['iş', 'acil'] };
    expect(matchesFilter(workCard, multiTagFilter)).toBe(true);

    const missingTagFilter: CardFilterState = { ...DEFAULT_FILTERS, tags: ['iş', 'proje'] };
    expect(matchesFilter(workCard, missingTagFilter)).toBe(false);
  });

  it('alışkanlık kökenine göre filtreler', () => {
    const habitCard = sampleCard({ habitId: 'h1' });
    const manualCard = sampleCard({ habitId: null });

    const habitFilter: CardFilterState = { ...DEFAULT_FILTERS, habit: 'habit' };
    const manualFilter: CardFilterState = { ...DEFAULT_FILTERS, habit: 'manual' };

    expect(matchesFilter(habitCard, habitFilter)).toBe(true);
    expect(matchesFilter(manualCard, habitFilter)).toBe(false);

    expect(matchesFilter(habitCard, manualFilter)).toBe(false);
    expect(matchesFilter(manualCard, manualFilter)).toBe(true);
  });

  it('son tarihe ve gecikmiş duruma göre filtreler', () => {
    const now = new Date('2026-08-19T12:00:00Z');
    const overdueCard = sampleCard({
      deadlineAt: '2026-08-19T10:00:00Z',
      done: false,
    });
    const upcomingCard = sampleCard({
      deadlineAt: '2026-08-19T15:00:00Z',
      done: false,
    });
    const completedOverdueCard = sampleCard({
      deadlineAt: '2026-08-19T10:00:00Z',
      done: true,
    });
    const noDeadlineCard = sampleCard({ deadlineAt: null });

    const hasDeadlineFilter: CardFilterState = { ...DEFAULT_FILTERS, deadline: 'has_deadline' };
    expect(matchesFilter(overdueCard, hasDeadlineFilter)).toBe(true);
    expect(matchesFilter(upcomingCard, hasDeadlineFilter)).toBe(true);
    expect(matchesFilter(noDeadlineCard, hasDeadlineFilter)).toBe(false);

    const overdueFilter: CardFilterState = { ...DEFAULT_FILTERS, deadline: 'overdue' };
    expect(matchesFilter(overdueCard, overdueFilter, now)).toBe(true);
    expect(matchesFilter(upcomingCard, overdueFilter, now)).toBe(false);
    expect(matchesFilter(completedOverdueCard, overdueFilter, now)).toBe(false);
    expect(matchesFilter(noDeadlineCard, overdueFilter, now)).toBe(false);
  });

  it('renk seçimine göre filtreler', () => {
    const blueCard = sampleCard({ color: 'blue' });
    const redCard = sampleCard({ color: 'red' });

    const blueFilter: CardFilterState = { ...DEFAULT_FILTERS, color: 'blue' };
    expect(matchesFilter(blueCard, blueFilter)).toBe(true);
    expect(matchesFilter(redCard, blueFilter)).toBe(false);
  });

  it('aktif filtre sayısını doğru hesaplar', () => {
    const filters: CardFilterState = {
      status: 'todo',
      priority: 'high',
      tags: ['iş', 'acil'],
      habit: 'manual',
      deadline: 'overdue',
      color: 'green',
    };
    expect(countActiveFilters(filters)).toBe(7); // status(1) + priority(1) + tags(2) + habit(1) + deadline(1) + color(1)
    expect(hasActiveFilters(filters)).toBe(true);
  });
});

File: mobile/test/filters_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:planner/api/models.dart';
import 'package:planner/filters.dart';

PlannerCard _sampleCard({
  String id = 'c1',
  String day = '2026-08-19',
  String title = 'Test',
  String note = '',
  String? startTime,
  String? endTime,
  String color = 'blue',
  bool done = false,
  double sortIndex = 600,
  bool manualSort = false,
  String? habitId,
  List<ChecklistItem> checklist = const [],
  String priority = 'none',
  String? deadlineAt,
  List<String> tags = const [],
  List<int> reminders = const [],
  List<CardImage> images = const [],
  String updatedAt = '2026-08-19T08:00:00Z',
}) =>
    PlannerCard(
      id: id,
      day: day,
      title: title,
      note: note,
      startTime: startTime,
      endTime: endTime,
      color: color,
      done: done,
      sortIndex: sortIndex,
      manualSort: manualSort,
      habitId: habitId,
      checklist: checklist,
      priority: priority,
      deadlineAt: deadlineAt,
      tags: tags,
      reminders: reminders,
      images: images,
      updatedAt: updatedAt,
    );

void main() {
  group('CardFilterState tests', () {
    test('default filters match all cards', () {
      final card = _sampleCard();
      expect(CardFilterState.defaultFilters.matches(card), isTrue);
      expect(CardFilterState.defaultFilters.activeCount, equals(0));
      expect(CardFilterState.defaultFilters.hasActiveFilters, isFalse);
    });

    test('filters by status', () {
      final todoCard = _sampleCard(done: false);
      final doneCard = _sampleCard(done: true);

      const todoFilter = CardFilterState(status: FilterStatus.todo);
      const doneFilter = CardFilterState(status: FilterStatus.done);

      expect(todoFilter.matches(todoCard), isTrue);
      expect(todoFilter.matches(doneCard), isFalse);

      expect(doneFilter.matches(todoCard), isFalse);
      expect(doneFilter.matches(doneCard), isTrue);
    });

    test('filters by priority', () {
      final urgentCard = _sampleCard(priority: 'urgent');
      final lowCard = _sampleCard(priority: 'low');

      const urgentFilter = CardFilterState(priority: 'urgent');
      expect(urgentFilter.matches(urgentCard), isTrue);
      expect(urgentFilter.matches(lowCard), isFalse);
    });

    test('filters by tags case-insensitively', () {
      final tagged = _sampleCard(tags: ['İş', 'Proje']);
      final other = _sampleCard(tags: ['Kişisel']);

      const tagFilter = CardFilterState(tags: {'iş'});
      expect(tagFilter.matches(tagged), isTrue);
      expect(tagFilter.matches(other), isFalse);
    });

    test('filters by habit origin', () {
      final habitCard = _sampleCard(habitId: 'h1');
      final manualCard = _sampleCard(habitId: null);

      const habitFilter = CardFilterState(habit: FilterHabit.habit);
      const manualFilter = CardFilterState(habit: FilterHabit.manual);

      expect(habitFilter.matches(habitCard), isTrue);
      expect(habitFilter.matches(manualCard), isFalse);

      expect(manualFilter.matches(habitCard), isFalse);
      expect(manualFilter.matches(manualCard), isTrue);
    });

    test('filters by deadline and overdue', () {
      final now = DateTime.parse('2026-08-19T12:00:00Z');
      final overdueCard = _sampleCard(
        deadlineAt: '2026-08-19T10:00:00Z',
        done: false,
      );
      final upcomingCard = _sampleCard(
        deadlineAt: '2026-08-19T15:00:00Z',
        done: false,
      );
      final completedCard = _sampleCard(
        deadlineAt: '2026-08-19T10:00:00Z',
        done: true,
      );

      const hasDeadline = CardFilterState(deadline: FilterDeadline.hasDeadline);
      expect(hasDeadline.matches(overdueCard), isTrue);
      expect(hasDeadline.matches(upcomingCard), isTrue);

      const overdueFilter = CardFilterState(deadline: FilterDeadline.overdue);
      expect(overdueFilter.matches(overdueCard, now), isTrue);
      expect(overdueFilter.matches(upcomingCard, now), isFalse);
      expect(overdueFilter.matches(completedCard, now), isFalse);
    });

    test('filters by color', () {
      final redCard = _sampleCard(color: 'red');
      final blueCard = _sampleCard(color: 'blue');

      const redFilter = CardFilterState(color: 'red');
      expect(redFilter.matches(redCard), isTrue);
      expect(redFilter.matches(blueCard), isFalse);
    });
  });
}
